layout_2020: Check blocks for whether they will establish a formatting context

stshine opened the pull request:
<!-- Please describe your changes on the following line: -->
According to spec
https://drafts.csswg.org/css-overflow-3/#overflow-control, If the computed value of overflow on a block box is neither visible nor clip nor a combination thereof, it establishes an independent formatting context for its contents.


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

<!-- Either: -->
- [x] There are tests for these changes OR

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

bors-servo commented:
:hourglass: Trying commit c38df221755c9a1054f11ad91214ac6f40cb107e with merge 3bd616a359d2e14ae2d6bf26c9d18d2584433fab...

mrobinson commented on components/layout_2020/flow/construct.rs:
```diff
@@ -591,7 +591,9 @@ where
             Ok(contents) => match display_inside {
-                DisplayInside::Flow { is_list_item } => {
+                DisplayInside::Flow { is_list_item }
+                    if !info.style.get_box().overflow_x.is_scrollable() =>
```
I think this should be a new function on the `StyleExt` crate, because there are [quite a few properties](https://developer.mozilla.org/en-US/docs/Web/Guide/CSS/Block_formatting_context) that trigger the creation of a new formatting context and we should try to handle them all. In addition, I think this should check both `overflow_x` and `overflow_y`.

Loirooriol replied:
I agree with turning this into its own function.

Checking `overflow-y` isn't probably necessary since elements can't be a scroll container in one axis but not in the other. The style adjuster takes care of changing the other property.

However, not all elements with `overflow: scroll` end up being scroll containers (e.g. replaced elements). So maybe it would be better to check layout rather than the computed style. Not sure if it matters in practice, though.

stshine replied:
I'm not very sure what the function should have since establishing formatting context is quite a complex topic and this check only happens when the element is a block, or, have `DisplayOutside::Block` and `DisplayInside::Flow`. Some is handled in other places(floats, abs-pos elements, `<html>`), Some is other kind of formatting context(Flexbox), and Some are not actually implemented(tables, multicol, grid).

mrobinson replied:
> I'm not very sure what the function should have since establishing formatting context is quite a complex topic and this check only happens when the element is a block, or, have `DisplayOutside::Block` and `DisplayInside::Flow`. Some is handled in other places(floats, abs-pos elements, `<html>`), Some is other kind of formatting context(Flexbox), and Some are not actually implemented(tables, multicol, grid).

It's true that not all situations would be useful to put in this function, but offhand this could also check whether the contents are `display: flow-root` when we add support for that as well as `column-span: all`. In short, I think there will need to be more situations in the not too distant future.

stshine replied:
> However, not all elements with `overflow: scroll` end up being scroll containers (e.g. replaced elements). So maybe it would be better to check layout rather than the computed style. Not sure if it matters in practice, though.

Since spec says bfc is established when `overflow` is not `visible` or `clip`, IMO this is OK.

Loirooriol replied:
Ah, that's right, it checks the compute value.

mrobinson reviewed (changes requested):
Thanks for doing this. We recently noticed this problem working on floats. A couple comments below:

bors-servo commented:
:broken_heart: Test failed - [checks-github](https://github.com/servo/servo/actions/runs/5122066069/jobs/9211529254)

stshine committed:
style: Enable 'column-span' for layout 2020

stshine committed:
layout_2020: Add 'establishes_block_formatting_context' method to 'ComputedValuesExt'

This method checks whether the style of a normal block would establish
a block formatting context.

stshine committed:
layout_2020: Check blocks for whether they will establish a formatting context

stshine commented:
Updated according to comments.

stshine commented:
Updated according to comments.

mrobinson commented on components/layout_2020/style_ext.rs:
```diff
@@ -365,6 +367,25 @@ impl ComputedValuesExt for ComputedValues {
     }
 
+    /// Return true if this style is a normal block and establishes
+    /// a new blockformatting context.
+    fn block_establish_formatting_context(&self) -> bool {
```
```suggestion
    /// a new block formatting context.
    fn establishes_block_formatting_context(&self) -> bool {
```

Let's keep the name consistent with the other methods here.
Nit: `blockformatting` →`block formatting`

mrobinson reviewed (approved):
Thank you! This looks great, other than 2 very small naming nits.

bors-servo commented:
:pushpin: Commit 8b0b849 has been approved by `mrobinson`

<!-- @bors-servo r=mrobinson 8b0b84948ef4c641095613f7517ad8caf571be6f -->

bors-servo commented:
:hourglass: Testing commit 8b0b84948ef4c641095613f7517ad8caf571be6f with merge ada8f1e93c5bd3270a15fe0f643bb692b88de9f0...

stshine commented:
oh wait a minute, haven't updated test expectation yet. investigating the failed tests :)

mrobinson commented:
> oh wait a minute, haven't updated test expectation yet. investigating the failed tests :)

Whoops sorry. I'll cancel the job.

bors-servo commented:
:broken_heart: Test failed - [checks-github](https://github.com/servo/servo/actions/runs/5134185846/jobs/9238607083)

stshine commented:
Looks like the tests failed because the following not implemented:

UAs must apply the [overflow](https://drafts.csswg.org/css2/#propdef-overflow) property set on the root element to the viewport. When the root element is an HTML "HTML" element or an XHTML "html" element, and that element has an HTML "BODY" element or an XHTML "body" element as a child, user agents must instead apply the overflow property from the first such child element to the viewport, if the value on the root element is [visible](https://drafts.csswg.org/css2/#valdef-overflow-visible). The visible value when used for the viewport must be interpreted as [auto](https://drafts.csswg.org/css2/#valdef-overflow-auto). The element from which the value is propagated must have a used value for overflow of visible.

Let's deal with that later.

stshine committed:
Update test expectations

mrobinson commented:
A fix for the above issue is on my todo list.

mrobinson commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 0806767 has been approved by `mrobinson`

<!-- @bors-servo r=mrobinson 08067674f4117d5dc422d3c4b7e23e01ffdda4ca -->

bors-servo commented:
:hourglass: Testing commit 08067674f4117d5dc422d3c4b7e23e01ffdda4ca with merge 6fa314de2a507befbc9b016aca631e2e09dc17f9...

bors-servo commented:
:sunny: Test successful - [checks-github](https://github.com/servo/servo/actions/runs/5135319118/jobs/9243402049)
Approved by: mrobinson
Pushing 6fa314de2a507befbc9b016aca631e2e09dc17f9 to master...
